File: backend/nfe/src/model/nfe_emitente_model.py
from src import db


class NfeEmitenteModel(db.Model):
    __tablename__ = 'nfe_emitente'

    id = db.Column(db.Integer, primary_key=True)
    cnpj = db.Column(db.String(14))
    cpf = db.Column(db.String(11))
    nome = db.Column(db.String(60))
    fantasia = db.Column(db.String(60))
    logradouro = db.Column(db.String(60))
    numero = db.Column(db.String(60))
    complemento = db.Column(db.String(60))
    bairro = db.Column(db.String(60))
    codigo_municipio = db.Column(db.Integer)
    nome_municipio = db.Column(db.String(60))
    uf = db.Column(db.String(2))
    cep = db.Column(db.String(8))
    codigo_pais = db.Column(db.Integer)
    nome_pais = db.Column(db.String(60))
    telefone = db.Column(db.String(14))
    inscricao_estadual = db.Column(db.String(14))
    inscricao_estadual_st = db.Column(db.String(14))
    inscricao_municipal = db.Column(db.String(15))
    cnae = db.Column(db.String(7))
    crt = db.Column(db.String(1))
    id_nfe_cabecalho = db.Column(db.Integer, db.ForeignKey('nfe_cabecalho.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_cabecalho = data.get('idNfeCabecalho')
        self.cnpj = data.get('cnpj')
        self.cpf = data.get('cpf')
        self.nome = data.get('nome')
        self.fantasia = data.get('fantasia')
        self.logradouro = data.get('logradouro')
        self.numero = data.get('numero')
        self.complemento = data.get('complemento')
        self.bairro = data.get('bairro')
        self.codigo_municipio = data.get('codigoMunicipio')
        self.nome_municipio = data.get('nomeMunicipio')
        self.uf = data.get('uf')
        self.cep = data.get('cep')
        self.codigo_pais = data.get('codigoPais')
        self.nome_pais = data.get('nomePais')
        self.telefone = data.get('telefone')
        self.inscricao_estadual = data.get('inscricaoEstadual')
        self.inscricao_estadual_st = data.get('inscricaoEstadualSt')
        self.inscricao_municipal = data.get('inscricaoMunicipal')
        self.cnae = data.get('cnae')
        self.crt = data.get('crt')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeCabecalho': self.id_nfe_cabecalho,
            'cnpj': self.cnpj,
            'cpf': self.cpf,
            'nome': self.nome,
            'fantasia': self.fantasia,
            'logradouro': self.logradouro,
            'numero': self.numero,
            'complemento': self.complemento,
            'bairro': self.bairro,
            'codigoMunicipio': self.codigo_municipio,
            'nomeMunicipio': self.nome_municipio,
            'uf': self.uf,
            'cep': self.cep,
            'codigoPais': self.codigo_pais,
            'nomePais': self.nome_pais,
            'telefone': self.telefone,
            'inscricaoEstadual': self.inscricao_estadual,
            'inscricaoEstadualSt': self.inscricao_estadual_st,
            'inscricaoMunicipal': self.inscricao_municipal,
            'cnae': self.cnae,
            'crt': self.crt,
        }

File: frontend/nfe/lib/app/data/model/nfe_emitente_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/infra/infra_imports.dart';
import 'package:nfe/app/data/domain/domain_imports.dart';

class NfeEmitenteModel {
	int? id;
	int? idNfeCabecalho;
	String? cnpj;
	String? cpf;
	String? nome;
	String? fantasia;
	String? logradouro;
	String? numero;
	String? complemento;
	String? bairro;
	int? codigoMunicipio;
	String? nomeMunicipio;
	String? uf;
	String? cep;
	int? codigoPais;
	String? nomePais;
	String? telefone;
	String? inscricaoEstadual;
	String? inscricaoEstadualSt;
	String? inscricaoMunicipal;
	String? cnae;
	String? crt;

	NfeEmitenteModel({
		this.id,
		this.idNfeCabecalho,
		this.cnpj,
		this.cpf,
		this.nome,
		this.fantasia,
		this.logradouro,
		this.numero,
		this.complemento,
		this.bairro,
		this.codigoMunicipio,
		this.nomeMunicipio,
		this.uf,
		this.cep,
		this.codigoPais,
		this.nomePais,
		this.telefone,
		this.inscricaoEstadual,
		this.inscricaoEstadualSt,
		this.inscricaoMunicipal,
		this.cnae,
		this.crt,
	});

	static List<String> dbColumns = <String>[
		'id',
		'cnpj',
		'cpf',
		'nome',
		'fantasia',
		'logradouro',
		'numero',
		'complemento',
		'bairro',
		'codigo_municipio',
		'nome_municipio',
		'uf',
		'cep',
		'codigo_pais',
		'nome_pais',
		'telefone',
		'inscricao_estadual',
		'inscricao_estadual_st',
		'inscricao_municipal',
		'cnae',
		'crt',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Cnpj',
		'Cpf',
		'Nome',
		'Fantasia',
		'Logradouro',
		'Numero',
		'Complemento',
		'Bairro',
		'Codigo Municipio',
		'Nome Municipio',
		'Uf',
		'Cep',
		'Codigo Pais',
		'Nome Pais',
		'Telefone',
		'Inscricao Estadual',
		'Inscricao Estadual St',
		'Inscricao Municipal',
		'Cnae',
		'Crt',
	];

	NfeEmitenteModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeCabecalho = jsonData['idNfeCabecalho'];
		cnpj = jsonData['cnpj'];
		cpf = jsonData['cpf'];
		nome = jsonData['nome'];
		fantasia = jsonData['fantasia'];
		logradouro = jsonData['logradouro'];
		numero = jsonData['numero'];
		complemento = jsonData['complemento'];
		bairro = jsonData['bairro'];
		codigoMunicipio = jsonData['codigoMunicipio'];
		nomeMunicipio = jsonData['nomeMunicipio'];
		uf = NfeEmitenteDomain.getUf(jsonData['uf']);
		cep = jsonData['cep'];
		codigoPais = jsonData['codigoPais'];
		nomePais = jsonData['nomePais'];
		telefone = jsonData['telefone'];
		inscricaoEstadual = jsonData['inscricaoEstadual'];
		inscricaoEstadualSt = jsonData['inscricaoEstadualSt'];
		inscricaoMunicipal = jsonData['inscricaoMunicipal'];
		cnae = jsonData['cnae'];
		crt = NfeEmitenteDomain.getCrt(jsonData['crt']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeCabecalho'] = idNfeCabecalho != 0 ? idNfeCabecalho : null;
		jsonData['cnpj'] = Util.removeMask(cnpj);
		jsonData['cpf'] = Util.removeMask(cpf);
		jsonData['nome'] = nome;
		jsonData['fantasia'] = fantasia;
		jsonData['logradouro'] = logradouro;
		jsonData['numero'] = numero;
		jsonData['complemento'] = complemento;
		jsonData['bairro'] = bairro;
		jsonData['codigoMunicipio'] = codigoMunicipio;
		jsonData['nomeMunicipio'] = nomeMunicipio;
		jsonData['uf'] = NfeEmitenteDomain.setUf(uf);
		jsonData['cep'] = Util.removeMask(cep);
		jsonData['codigoPais'] = codigoPais;
		jsonData['nomePais'] = nomePais;
		jsonData['telefone'] = telefone;
		jsonData['inscricaoEstadual'] = inscricaoEstadual;
		jsonData['inscricaoEstadualSt'] = inscricaoEstadualSt;
		jsonData['inscricaoMunicipal'] = inscricaoMunicipal;
		jsonData['cnae'] = cnae;
		jsonData['crt'] = NfeEmitenteDomain.setCrt(crt);
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeCabecalho = plutoRow.cells['idNfeCabecalho']?.value;
		cnpj = plutoRow.cells['cnpj']?.value;
		cpf = plutoRow.cells['cpf']?.value;
		nome = plutoRow.cells['nome']?.value;
		fantasia = plutoRow.cells['fantasia']?.value;
		logradouro = plutoRow.cells['logradouro']?.value;
		numero = plutoRow.cells['numero']?.value;
		complemento = plutoRow.cells['complemento']?.value;
		bairro = plutoRow.cells['bairro']?.value;
		codigoMunicipio = plutoRow.cells['codigoMunicipio']?.value;
		nomeMunicipio = plutoRow.cells['nomeMunicipio']?.value;
		uf = plutoRow.cells['uf']?.value != '' ? plutoRow.cells['uf']?.value : 'AC';
		cep = plutoRow.cells['cep']?.value;
		codigoPais = plutoRow.cells['codigoPais']?.value;
		nomePais = plutoRow.cells['nomePais']?.value;
		telefone = plutoRow.cells['telefone']?.value;
		inscricaoEstadual = plutoRow.cells['inscricaoEstadual']?.value;
		inscricaoEstadualSt = plutoRow.cells['inscricaoEstadualSt']?.value;
		inscricaoMunicipal = plutoRow.cells['inscricaoMunicipal']?.value;
		cnae = plutoRow.cells['cnae']?.value;
		crt = plutoRow.cells['crt']?.value != '' ? plutoRow.cells['crt']?.value : 'AAA';
	}	

	NfeEmitenteModel clone() {
		return NfeEmitenteModel(
			id: id,
			idNfeCabecalho: idNfeCabecalho,
			cnpj: cnpj,
			cpf: cpf,
			nome: nome,
			fantasia: fantasia,
			logradouro: logradouro,
			numero: numero,
			complemento: complemento,
			bairro: bairro,
			codigoMunicipio: codigoMunicipio,
			nomeMunicipio: nomeMunicipio,
			uf: uf,
			cep: cep,
			codigoPais: codigoPais,
			nomePais: nomePais,
			telefone: telefone,
			inscricaoEstadual: inscricaoEstadual,
			inscricaoEstadualSt: inscricaoEstadualSt,
			inscricaoMunicipal: inscricaoMunicipal,
			cnae: cnae,
			crt: crt,
		);			
	}

	
}